feat: Add support for Permissioned Keys

konichuvak opened the pull request:
Adds functionality to enable signing transactions using the authenticator flow on the protocol.

Basically following the footsteps of #317 to add the same functionality to the Python client.

Commit:
fix: bumps GAS_MULTIPLIER to 1.7 to allow authenticator txs

Commit:
feat: adds support for authentication with permissioned key

- bumps v4-proto and grpcio to get protobufs for authenticators
- adds client methods for adding / removing / retrieving authenticators
- adds TxOption parameter to place / cancel methods to allow signing orders on behalf of another wallet
- modifies transaction building flow to include TxOptions

Commit:
feat: adds e2e example for using permissioned keys for placing orders

pnowosie commented on v4-client-py-v2/examples/permissioned_keys_example.py:
I was not able to run this example. Here is the error message:
```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.NOT_FOUND
	details = "account dydx18sukah44zfkjndlhcdmhkjnarl2sevhwf894vh not found"
```

Account `dydx18sukah44zfkjndlhcdmhkjnarl2sevhwf894vh` doesn't exists on testnet.
Have you succeeded running the example?

pnowosie replied:
Additionally, I would like to see the output when running the example. Could you introduce some print statements? e.g. [examples/batch_cancel_example.py](https://github.com/dydxprotocol/v4-clients/blob/74e45e6b404a435bf71e8ab741b6e58d499a831a/v4-client-py-v2/examples/batch_cancel_example.py)

konichuvak replied:
I am able to run the example with my own accounts on the mainnet.
The example there is just a copy of the JS client example, which does not run for me either btw.

I can parametrize the example such that the user would have to supply their own mainnet keys. Let me know if you have better ideas.

pnowosie replied:
I was able to fix the example and run it on testnet.
The reason it might work for you on mainnet is because you used owner's wallet to sign so no authenticator was needed 
https://github.com/dydxprotocol/v4-clients/blob/735cb30dda8438a0b73571d0d63afbde102a8530/v4-client-py-v2/examples/permissioned_keys_example.py#L76-L78

konichuvak replied:
glad you made it work in the end!

good, point! that should be a `trader_wallet` there

pnowosie commented on v4-client-py-v2/dydx_v4_client/node/authenticators.py:
```diff
@@ -0,0 +1,137 @@
+    return base64.b64decode(value)
+
+
+def decode_authenticator(config: str, auth_type: str) -> Union[SubAuthenticator, Authenticator]:
```
Is this function used anywhere?
I'd prefer not to introduce redundant code.

pnowosie reviewed (commented):
@konichuvak Thank you very much for you contribution. 🙌 

I'd like to kindly ask you to check the comments below and respond if you think they're make sense.

Also I spotted a formatting issues. Please make sure you run `poetry run black .` to reformat the code before pushing.

Thanks 🙏

pnowosie commented:
@konichuvak Please ignore my above comments. 
We've decided to implement this feature and I will be happy to use your code and modify to our needs.
I will open an subsequent PR and mention this one.

konichuvak commented on v4-client-py-v2/dydx_v4_client/network.py:
```diff
@@ -56,6 +55,21 @@ def make_config(
 make_mainnet = partial(make_secure, mainnet_node)
 
 
+staging_node = partial(
+    NodeConfig,
+    "dydxprotocol-testnet",
+    chaintoken_denom="adv4tnt",
+    usdc_denom="ibc/8E27BA2D5493AF5636760E354E46004562C46AB7EC0CC4C1CA14E9E20E2545B5",
+)
+make_staging = partial(
+    make_secure,
+    staging_node,
+    rest_indexer="https://indexer.v4staging.dydx.exchange/",
+    websocket_indexer="wss://indexer.v4staging.dydx.exchange/v4/ws",
+    node_url="validator.v4staging.dydx.exchange",
+)
+STAGING = make_staging()
```
I tried to make it similar to the JS client impl, since they are using staging there for the e2e example.
But as you pointed in the comment below one of the addresses that is used there does not exist though, so this does not serve much purpose indeed.

Commit:
fix: removes unused staging node setup

Commit:
fix: addresses comments in review

- runs linter
- replaces TypedDict with dataclass
- replaces starred imports with explicit ones

konichuvak commented:
> @konichuvak Please ignore my above comments. We've decided to implement this feature and I will be happy to use your code and modify to our needs. I will open an subsequent PR and mention this one.

No worries, hope you find it useful to have as a reference in your implementation.

Commit:
chore: bump `v4-client-py-v2` version to `1.1.5`

pnowosie commented:
> No worries, hope you find it useful to have as a reference in your implementation.

I find it very useful, I will make sure to mention your contribution in my PR
